### PR TITLE
Tests/bootstrap: Add support to read `WP_TESTS_PHPUNIT_POLYFILLS_PATH` from environment variables

### PR DESCRIPTION
**Pull request by shivammathur**
This PR adds support to read the WP_TESTS_PHPUNIT_POLYFILLS_PATH from an environment variable.

Currently one has to define the `WP_TESTS_PHPUNIT_POLYFILLS_PATH` as a PHP constant.

Adding support for reading this from an environment variable would allow setup-php GitHub Action to set this variable.

Trac ticket: https://core.trac.wordpress.org/ticket/54470

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by shivammathur** [comment]
@jrfnl 

Please review this PR, once merged I can set `WP_TESTS_PHPUNIT_POLYFILLS_PATH` when `phpunit-polyfills` is installed in setup-php, Then it will be easier for users to use their own PHPUnit in tools input of setup-php.

File: tests/phpunit/includes/bootstrap.php [hunk 1]
```diff
@@ -82,6 +82,12 @@
 	$phpunit_polyfills_autoloader = dirname( dirname( dirname( __DIR__ ) ) ) . '/vendor/yoast/phpunit-polyfills/phpunitpolyfills-autoload.php';
 	$phpunit_polyfills_error      = false;
 
+	// Define WP_TESTS_PHPUNIT_POLYFILLS_PATH if set as environment variable.
```
Comment (jrfnl) [inline tests/phpunit/includes/bootstrap.php:85]: If this gets added, the option to set the path via an environment variable should also be mentioned in the multi-line comment above this conditional code-block.

File: tests/phpunit/includes/bootstrap.php [hunk 2]
```diff
@@ -82,6 +82,12 @@
...
+	$phpunit_polyfills_path = getenv( 'WP_TESTS_PHPUNIT_POLYFILLS_PATH' );
+	if ( $phpunit_polyfills_path ) {
+		define( 'WP_TESTS_PHPUNIT_POLYFILLS_PATH', $phpunit_polyfills_path );
```
Comment (jrfnl) [inline tests/phpunit/includes/bootstrap.php:88]: Could we use a different variable name for this ? Maybe `$phpunit_polyfills_path_env`  ?
I'm a bit wary of re-using the same variable name for two scenarios.

**Review by jrfnl — COMMENTED** [review]
@shivammathur Just left some notes to think over.

**Comment by jrfnl** [comment]
> Then it will be easier for users to use their own PHPUnit in tools input of setup-php.

The thing is: users **shouldn't** use their own PHPUnit via `tools` in `setup-php`.

There's two reasons for that:
1. Performance: they will be installing PHPUnit twice in that case, once as a PHAR via `tools` and once via `composer global` as it automatically gets installed with the PHPUnit Polyfills.
2. The Polyfills are designed to be used with the most appropriate PHPUnit version for the PHP version tests are being run on. When a user requires a specific PHPUnit PHAR via `tools`, they are likely to create a mismatch, which can cause all sorts of problems when running the tests.

I just saw issue shivammathur/setup-php/#531 in the `setup-php` repo. Is this PR in response to that ticket ? Would you like me to respond to the user in that ticket ?

**Comment by shivammathur** [comment]
@jrfnl 
Yes, I understand that the most optimal way would be to add `phpunit-polyfills` and remove `phpunit` from tools input and use the PHPUnit version that `phpunit-polyfills` sets.

But there are existing working workflows that set the PHPUnit versions in the matrix and it would be easier for them to just add `phpunit-polyfills` to the tools input and it just works again.
For example the workflow in https://github.com/shivammathur/setup-php/issues/529 was breaking when running with  `phpunit-polyfills`'s PHPUnit versions and most likely would have required refactoring the tests.

So this would make my life easier as it would give people the choice.

File: tests/phpunit/includes/bootstrap.php (hunk 1 above)
Comment (shivammathur) [reply tests/phpunit/includes/bootstrap.php:85]: Added

File: tests/phpunit/includes/bootstrap.php (hunk 2 above)
Comment (shivammathur) [reply tests/phpunit/includes/bootstrap.php]: Done

**Comment by shivammathur** [comment]
@jrfnl 
Let me know what you think.
If this will increase the maintenance burden on your end due to version mismatches, I will close this.

**Comment by jrfnl** [comment]
> But there are existing working workflows that set the PHPUnit versions in the matrix and it would be easier for them to just add `phpunit-polyfills` to the tools input and it just works again.

Except that's not true.

> For example the workflow in [shivammathur/setup-php#529](https://github.com/shivammathur/setup-php/issues/529) was breaking when running with `phpunit-polyfills`'s PHPUnit versions and most likely would have required refactoring the tests.

If they continue to run their integration tests - WITHOUT making the required changes as per the [WP Make Core](https://make.wordpress.org/core/2021/09/27/changes-to-the-wordpress-core-php-test-suite/) blogpost -, their tests are **_inherently_** broken now as the typical fixtures methods which they inherit from the WP test suite will no longer run when testing against WP 5.9/`trunk`/`master`, UNLESS they make the changes.

The changes aren't huge or anything, they are actually quite straight-forward to make, but they MUST be made for their integration tests to continue working.

This is inherit to the choice they made to base their tests on the WP native test suite, which means that if the WP Core test suite changes, their tests will also need to be updated.

While they may not see any tests breaking now, they will soon enough, once they start testing against WP 5.9, or else, they may start running into weird race-conditions in their tests due to the fixtures no longer running.

> So this would make my life easier as it would give people the ability to choose the PHPUnit version.

I understand and appreciate that, but you shouldn't be the person who should need to solve their workflow problems, when they are not following the very clear instructions which have been publicly posted for them to follow for over two months.

IMO, point them to the blogpost on Make and tell them to follow the instructions given there or, if needs be, ping me in an issue and I will put them right.

I very much realize that plugins/themes which want to test against a range of WP versions need to make some accommodations in their workflows for the PHPUnit versions which WP allows for (which have only now been widened in WP 5.9 to encompass the full range), but there are other ways to do that.

**Comment by shivammathur** [comment]
>If they continue to run their integration tests - WITHOUT making the required changes as per the WP Make Core blogpost -, their tests are inherently broken now as the typical fixtures methods which they inherit from the WP test suite will no longer run when testing against WP 5.9/trunk/master, UNLESS they make the changes.

Thanks for the detailed response, I wasn't aware of this.
From now on, I will point to the blog post for any issues regarding this. Closing the PR.

**Comment by jrfnl** [comment]
> @jrfnl Let me know what you think. If this will increase the maintenance burden on your end due to version mismatches, I will close this.

@shivammathur It's not so much about the maintenance burden as it is about people _doing it wrong_ and this solution would allow them to continue to _do it wrong_.

What the best way to solve it is, depends on the WP versions they want to support/test against:
* If only testing against the current `develop` version or WP 5.9 and higher: by all means, they can safely use `tools: phpunit-polyfills` and things will just work.
* If they also want to test against older WP versions - WP 5.2 - 5.8 -, they will need to install an appropriate PHPUnit version within the PHPUnit 5.x-7.x range specifically for those builds and should not try to run the tests against PHP 8.1.
    The _recommended_ way to set up their tests in that case is to use a project local Composer install with PHPUnit + the polyfills and a PHPUnit version constraint for 5.x-7x and to on-the-fly remove the PHPUnit require for running against WP 5.9+ (the polyfills will then kick in and install the more appropriate PHPUnit version).
    They still _could_ use the `tools: phpunit-polyfills` if they insist. In that case, they should run a `composer global phpunit/phpunit:"${{ matrix.phpunit-version}}" command to switch the PHPUnit version (as PHPUnit will be globally installed via Composer already).
* If they also still want to test against yet older WP version, things get even more complicated and I'd recommend for them to use the WP Test Utils (also mentioned in the blogpost) as there are some more complications to get round in that case.
    Other than that, the version switching as described above would still apply.

P.S.: just noticed you closed the PR while I was still writing this. Please also feel free to point people to this PR as I imagine the explanations I've given here may help them as well.

**Pull request closed without merging**